sparse_ho/criterion.py

```diff
@@ -355,76 +308,69 @@ def __init__(self, X, y, model, sigma, finite_difference_step=None,
         # TODO add warm start
         mask, dense, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y, log_alpha, self.model,
+            X[self.idx_train], y[self.idx_train], log_alpha, model,
             tol=tol, mask0=self.mask0, dense0=self.dense0, compute_jac=False)
         mask2, dense2, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y + self.epsilon * self.delta,
-            log_alpha, self.model,
-            tol=tol, compute_jac=False)
+            X[self.idx_train], y[self.idx_train] + self.epsilon * self.delta,
+            log_alpha, model, tol=tol, compute_jac=False)
 
-        val = self.value(mask, dense, mask2, dense2)
+        val = self.value_outer_crit(mask, dense, mask2, dense2)
 
         return val
 
+    def _init_delta_epsilon(self, X):
+        if self.finite_difference_step:
+            self.epsilon = self.finite_difference_step
+        else:
+            # Use Deledalle et al. 2014 heuristic
+            self.epsilon = 2.0 * self.sigma / (X.shape[0]) ** 0.3
+        rng = check_random_state(self.random_state)
+        self.delta = rng.randn(X.shape[0])  # sample random noise for MCMC step
+        self.init_delta_epsilon = True
+
     def get_val_grad(
-            self, log_alpha, get_beta_jac_v,
+            self, model, X, y, log_alpha, get_beta_jac_v,
             mask0=None, dense0=None, beta_star=None,
             jac0=None, max_iter=1000, tol=1e-3, compute_jac=True):
+        if not self.init_delta_epsilon:
+            self._init_delta_epsilon(X)
+
+        def v(mask, dense):
+            return (2 * X[:, mask].T @ (
+                    X[:, mask] @ dense - y -
+                    self.delta * self.sigma ** 2 / self.epsilon))
+
+        def v2(mask, dense):
+            return ((2 * self.sigma ** 2 *
+                     X[:, mask].T @ self.delta / self.epsilon))
+
         mask, dense, jac_v, quantity_to_warm_start = get_beta_jac_v(
-            self.model.X, self.model.y, log_alpha, self.model, self.v,
+            X, y, log_alpha, model, v,
             mask0=self.mask0, dense0=self.dense0,
             quantity_to_warm_start=self.quantity_to_warm_start,
             max_iter=max_iter, tol=tol, compute_jac=compute_jac,
             full_jac_v=True)
         mask2, dense2, jac_v2, quantity_to_warm_start2 = get_beta_jac_v(
-            self.model.X, self.model.y + self.epsilon * self.delta,
-            log_alpha, self.model, self.v2, mask0=self.mask02,
+            X, y + self.epsilon * self.delta,
+            log_alpha, model, v2, mask0=self.mask02,
             dense0=self.dense02,
             quantity_to_warm_start=self.quantity_to_warm_start2,
             max_iter=max_iter, tol=tol, compute_jac=compute_jac,
             full_jac_v=True)
-        val = self.value(mask, dense, mask2, dense2)
-        self.value_test(mask, dense)
-        self.compute_rmse(mask, dense, beta_star)
+        val = self.value_outer_crit(X, y, mask, dense, mask2, dense2)
```

Comment:
```suggestion
        val = self.get_val_outer(X, y, mask, dense, mask2, dense2)
```

so we have : get_val, get_val_grad for inner and get_val_outer ?

Reply:
:+1: and even get_val_inner, get_val_outer to clarify

Reply:
> get_val, get_val_grad for inner and get_val_outer ?

No, ```get_val_grad``` computes the value and the gradient of the outer criterion given a value of the regularization parameter, ```get_val``` compute the value  of the outer criterion given the value of the regularization parameter.

```get_val_outer``` computes the value of the outer criterion, given the coefficients.

If you think of clearer names I am in!

Reply:
for starters get_val_and_grad for me, since I read get_val_grad as get_value_gradient, i.e. get gradient of value (and not, get value and its gradient)